src/control/lib/control/auto_test.go

```diff
@@ -519,25 +549,126 @@ func TestControl_AutoConfig_checkNetwork(t *testing.T) {
 
-			_, gotCheckErr := req.checkNetwork(context.Background(), resp)
+			diff := cmp.Diff(expResp.GetHostErrors(), resp.GetHostErrors(), cmpOpts...)
+			if diff != "" {
+				t.Fatalf("unexpected response (-want, +got):\n%s\n", diff)
+			}
 
-			if diff := cmp.Diff(expResp.GetHostErrors(), resp.GetHostErrors(), cmpOpts...); diff != "" {
+			if diff := cmp.Diff(expResp, resp, cmpOpts...); diff != "" {
 				t.Fatalf("unexpected response (-want, +got):\n%s\n", diff)
 			}
+		})
+	}
+}
+
+func TestControl_AutoConfig_checkCpus(t *testing.T) {
+	errCores := func(ssds, want, got int) error {
+		return errors.Errorf(errInsufNumCores, ssds, want, got)
+	}
+
+	for name, tc := range map[string]struct {
+		numCores      int
+		bdevCounts    []int // one for each I/O Server in resp.ConfigOut
+		expTgtCounts  []int
+		expHlprCounts []int
+		expCheckErr   error
+	}{
+		"no cores":           {expCheckErr: errors.Errorf(errInvalNumCores, 0)},
+		"24 cores no ssds":   {24, []int{0}, []int{16}, []int{7}, nil},
+		"24 cores 1 ssds":    {24, []int{1}, []int{16}, []int{7}, nil},
+		"24 cores 2 ssds":    {24, []int{2}, []int{16}, []int{7}, nil},
+		"24 cores 3 ssds":    {24, []int{3}, []int{12}, []int{11}, nil},
+		"24 cores 4 ssds":    {24, []int{4}, []int{16}, []int{7}, nil},
+		"24 cores 5 ssds":    {24, []int{5}, []int{15}, []int{8}, nil},
+		"24 cores 8 ssds":    {24, []int{8}, []int{16}, []int{7}, nil},
+		"24 cores 9 ssds":    {24, []int{9}, []int{18}, []int{5}, nil},
+		"24 cores 10 ssds":   {24, []int{10}, []int{20}, []int{3}, nil},
+		"24 cores 16 ssds":   {24, []int{16}, []int{16}, []int{7}, nil},
+		"18 cores no ssds":   {18, []int{0}, []int{16}, []int{1}, nil},
+		"18 cores 1 ssds":    {18, []int{1}, []int{16}, []int{1}, nil},
+		"18 cores 2 ssds":    {18, []int{2}, []int{16}, []int{1}, nil},
+		"18 cores 3 ssds":    {18, []int{3}, []int{12}, []int{5}, nil},
+		"18 cores 4 ssds":    {18, []int{4}, []int{16}, []int{1}, nil},
+		"18 cores 5 ssds":    {18, []int{5}, []int{15}, []int{2}, nil},
+		"18 cores 8 ssds":    {18, []int{8}, []int{16}, []int{1}, nil},
+		"18 cores 9 ssds":    {18, []int{9}, []int{9}, []int{8}, nil},
+		"18 cores 10 ssds":   {18, []int{10}, []int{10}, []int{7}, nil},
+		"18 cores 16 ssds":   {18, []int{16}, []int{16}, []int{1}, nil},
+		"16 cores no ssds":   {16, []int{0}, []int{16}, []int{1}, errCores(0, 17, 16)},
+		"16 cores 1 ssds":    {16, []int{1}, []int{16}, []int{1}, errCores(1, 17, 16)},
+		"16 cores 2 ssds":    {16, []int{2}, []int{16}, []int{1}, errCores(2, 17, 16)},
+		"16 cores 3 ssds":    {16, []int{3}, []int{12}, []int{3}, nil},
+		"16 cores 4 ssds":    {16, []int{4}, []int{16}, []int{1}, errCores(4, 17, 16)},
+		"16 cores 5 ssds":    {16, []int{5}, []int{15}, []int{0}, nil},
+		"16 cores 6 ssds":    {16, []int{6}, []int{12}, []int{3}, nil},
+		"16 cores 7 ssds":    {16, []int{7}, []int{14}, []int{1}, nil},
+		"16 cores 8 ssds":    {16, []int{8}, []int{16}, []int{1}, errCores(8, 17, 16)},
```

Comment:
These failure cases seem so odd to me, it makes it feel like folks with a 16 core CPU are out of luck. Couldn't we drop back to the next smallest multiple of the number of SSDs, even if it's less performant?

Reply:
I have been following guidance from @johannlombardi as documented in the ticket, it's expected that most installations will have sockets with 18 physical cores or more so the 16 and 32 core tests are edge cases, common use cases will be 18 & 24. Happy to adjust the selection if wanted.

Reply:
I'd assume Johann knows more about the use cases than I do. If it's possible to have 16 cores, though, my instinct would be to produce some kind of working config for them, even though it's not as performant. May not be a high priority though.